main/src/ca/uwaterloo/flix/language/phase/Deriver.scala

```diff
@@ -539,6 +540,138 @@ object Deriver extends Phase[KindedAst.Root, KindedAst.Root] {
+
+      val param = Symbol.freshVarSym("x", loc)
+      val exp = mkHashImpl(enum, param, loc, root)
+      val spec = mkHashSpec(enum, param, loc, root)
+
+      val defn = KindedAst.Def(hashDefSym, spec, exp)
+
+      val tconstrs = getTypeConstraintsForTypeParams(tparams, hashClassSym, loc)
+
+      KindedAst.Instance(
+        doc = Ast.Doc(Nil, loc),
+        mod = Ast.Modifiers.Empty,
+        sym = hashClassSym,
+        tpe = sc.base,
+        tconstrs = tconstrs,
+        defs = List(defn),
+        ns = Name.RootNS,
+        loc = loc
+      )
+  }
+
+  /**
+    * Creates the hash implementation for the given enum, where `param` is the parameter to the function.
+    */
+  private def mkHashImpl(enum: KindedAst.Enum, param: Symbol.VarSym, loc: SourceLocation, root: KindedAst.Root)(implicit flix: Flix): KindedAst.Expression = enum match {
+    case KindedAst.Enum(_, _, _, _, _, cases, _, _, _) =>
+      // create a match rule for each case
+      val matchRules = cases.values.zipWithIndex.map {
+        case (caze, index) => mkHashMatchRule(caze, index, loc, root)
+      }
+
+      // group the match rules in an expression
+      KindedAst.Expression.Match(
+        mkVarExpr(param, loc),
+        matchRules.toList,
+        loc
+      )
+  }
+
+  /**
+    * Creates the hash spec for the given enum, where `param` is the parameter to the function.
+    */
+  private def mkHashSpec(enum: KindedAst.Enum, param: Symbol.VarSym, loc: SourceLocation, root: KindedAst.Root)(implicit flix: Flix): KindedAst.Spec = enum match {
+    case KindedAst.Enum(_, _, _, tparams, _, _, _, sc, _) =>
+      val hashClassSym = PredefinedClasses.lookupClassSym("Hash", root)
+      KindedAst.Spec(
+        doc = Ast.Doc(Nil, loc),
+        ann = Nil,
+        mod = Ast.Modifiers.Empty,
+        tparams = tparams,
+        fparams = List(KindedAst.FormalParam(param, Ast.Modifiers.Empty, sc.base, loc)),
+        sc = Scheme(
+          tparams.map(_.tpe),
+          List(Ast.TypeConstraint(hashClassSym, sc.base, loc)),
+          Type.mkPureArrow(sc.base, Type.mkInt32(loc), loc)
+        ),
+        tpe = Type.mkInt32(loc),
+        eff = Type.Cst(TypeConstructor.True, loc),
+        loc = loc
+      )
+  }
+
+  /**
+    * Creates a ToString match rule for the given enum case.
+    */
+  private def mkHashMatchRule(caze: KindedAst.Case, index: Int, loc: SourceLocation, root: KindedAst.Root)(implicit flix: Flix): KindedAst.MatchRule = caze match {
+    case KindedAst.Case(_, _, _, sc) =>
+      val hashSigSym = PredefinedClasses.lookupSigSym("Hash", "hash", root)
+
+      // get a pattern corresponding to this case, e.g.
+      // `case C2(x0, x1)`
+      val (pat, varSyms) = mkPattern(sc.base, "x", loc)
+
+      val guard = KindedAst.Expression.True(loc)
+
+      // build a hash code by repeatedly adding the next hash and multiplying by 31
+      // the first hash is the index
+      // `((2 * 31) + hash(x0)) * 31 + hash(y0)`
```

Comment:
Is this the standard thing to do? I remember that multiplying by primes is supposedly good. 

I guess hashing is a tricky business in general.

Reply:
This is the implementation used in Java's `Arrays.hashCode`